pkg/controller/cnr/cnr_indicator.go

```diff
@@ -0,0 +1,209 @@
+			if numa.Type != v1alpha1.TopologyTypeNuma {
+				// only check numa
+				continue
+			}
+			numaCpuAllocatableSum += int(numa.Resources.Allocatable.Cpu().AsApproximateFloat64())
+			numaMemAllocatableSum += int(numa.Resources.Allocatable.Memory().AsApproximateFloat64())
+		}
+	}
+	klog.Infof("[CNRIndicatorNumaAllocatableSumAnomaly] numaCpuAllocatableSum: %d, numaMemAllocatableSum: %d", numaCpuAllocatableSum, numaMemAllocatableSum)
+	if numaCpuAllocatableSum != nodeCpuAllocatable || numaMemAllocatableSum != nodeMemAllocatable {
+		return true
+	}
+	return false
+}
+
+// checkPodAllocationSumAnomaly checks whether the numa's sum of pod allocation is greater than the numa allocatable
+func (ctrl *CNRController) checkPodAllocationSumAnomaly(cnr *v1alpha1.CustomNodeResource) bool {
+	qosConf := generic.NewQoSConfiguration()
+	for _, socket := range cnr.Status.TopologyZone {
+		for _, numa := range socket.Children {
+			if numa.Type != v1alpha1.TopologyTypeNuma {
+				// only check numa
+				continue
+			}
+			numaCpuAllocatable, numaMemAllocatable := int(numa.Resources.Allocatable.Cpu().AsApproximateFloat64()), int(numa.Resources.Allocatable.Memory().AsApproximateFloat64())
+			klog.Infof("[CNRIndicatorPodAllocationSumAnomaly] numaCpuAllocatable: %d, numaMemAllocatable: %d", numaCpuAllocatable, numaMemAllocatable)
+			podCpuAllocationSum, podMemAllocationSum := 0, 0
+			for _, allocation := range numa.Allocations {
+				key := allocation.Consumer
+				namespace, podname, _, err := native.ParseUniqObjectUIDKey(key)
+				if err != nil {
+					klog.Errorf("[CNRIndicatorPodAllocationSumAnomaly] failed to parse uniq object uid key %s", key)
+				}
+				pod, err := ctrl.client.KubeClient.CoreV1().Pods(namespace).Get(ctrl.ctx, podname, metav1.GetOptions{ResourceVersion: "0"})
+				if err != nil {
+					klog.Errorf("[CNRIndicatorPodAllocationSumAnomaly] failed to get pod %s", key)
+				}
+				// only check the pod with numa binding
+				if qos.IsPodNumaBinding(qosConf, pod) {
+					podCpuAllocationSum += int(allocation.Requests.Cpu().AsApproximateFloat64())
+					podMemAllocationSum += int(allocation.Requests.Memory().AsApproximateFloat64())
+				}
+			}
+			klog.Infof("[CNRIndicatorPodAllocationSumAnomaly] podCpuAllocationSum: %d, podMemAllocationSum: %d", podCpuAllocationSum, podMemAllocationSum)
+			if podCpuAllocationSum > numaCpuAllocatable || podMemAllocationSum > numaMemAllocatable {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// emitCNRAnomalyMetric emit CNR anomaly metric
+func (ctrl *CNRController) emitCNRAnomalyMetric(cnr *v1alpha1.CustomNodeResource, reason string) error {
+	node, err := ctrl.client.KubeClient.CoreV1().Nodes().Get(ctrl.ctx, cnr.Name, metav1.GetOptions{ResourceVersion: "0"})
+	if err != nil {
+		klog.Errorf("[CNREmitAnomalyMetric] failed to get node %s", cnr.Name)
+		return err
+	}
+
+	_ = ctrl.metricsEmitter.StoreInt64(metricsNameCNRReportAnomaly, 1, metrics.MetricTypeNameRaw,
+		metrics.MetricTag{
```

Comment:
why this metrics need node_ip tag?

Reply:
This allows us to know which cnr has the anomaly?

Reply:
the node_name has show which cnr is the abnormal, and add node_ip tag may be redundancy.

Reply:
Make sense, I will fix it.